File: packages/core/src/assets/asset_reference.rs
//! # Asset Reference
//!
//! The typed asset reference struct used everywhere in XACE.
//! Per Audit 2 — asset references are NEVER raw strings.
//! Always a typed AssetReference with a known status.
//!
//! ## Why Typed References
//! Raw strings break at runtime with no warning — a typo in an asset
//! path only fails when the engine tries to load it. Typed references
//! with explicit status (PLACEHOLDER/LINKED/MISSING/UNRESOLVED) make
//! asset state visible at the schema level, not just at load time.
//!
//! ## Auto-Naming Convention
//! Asset IDs follow a strict convention:
//! [entity_type]_[entity_name]_[asset_type]_v[N]
//! Examples:
//!   character_knight_mesh_v1
//!   enemy_dragon_roar_sfx_v1
//!   prop_chest_texture_v1
//!
//! This convention is enforced by the Asset Registry (Phase 7).
//! The AssetReference struct stores the ID — validation happens upstream.
//!
//! ## Global Invariant I12
//! UNRESOLVED asset references must never enter a committed CGS.
//! The Schema Factory validates this before every commit.
//! An UNRESOLVED reference means the asset was referenced in a schema
//! definition but never registered in the asset pipeline — this is a bug.

use serde::{Deserialize, Serialize};
use crate::assets::asset_type::AssetType;
use crate::assets::asset_status::AssetStatus;

// ── Asset Reference ───────────────────────────────────────────────────────────

/// A typed, tracked reference to an external asset.
///
/// Used everywhere an asset is referenced in XACE — component fields,
/// schema definitions, blueprint declarations. Never a raw string.
///
/// ## Lifecycle
/// UNRESOLVED → PLACEHOLDER → LINKED | MISSING
///
/// New entities start with PLACEHOLDER references auto-created by XACE.
/// The asset pipeline transitions them to LINKED when real files are mapped.
/// MISSING means a previously linked file can no longer be found.
/// UNRESOLVED blocks CGS commit entirely (I12).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AssetReference {
    /// Unique identifier for this asset.
    /// Follows auto-naming convention: [entity_type]_[name]_[type]_v[N]
    /// Assigned by the Asset Registry — never set manually.
    pub id: String,

    /// What kind of asset this reference points to.
    /// Determines how the engine adapter loads and uses the asset.
    pub asset_type: AssetType,

    /// Current state in the asset pipeline.
    /// Drives how the engine adapter handles this reference each tick.
    pub status: AssetStatus,
}

impl AssetReference {
    /// Creates a new asset reference in Placeholder state.
    ///
    /// This is the correct and only way to create a new asset reference.
    /// All new references start as Placeholder — never as Linked or Unresolved.
    /// The asset pipeline transitions status from Placeholder to Linked
    /// when a real file is mapped via the Asset Linker (Phase 7).
    pub fn placeholder(id: impl Into<String>, asset_type: AssetType) -> Self {
        Self {
            id: id.into(),
            asset_type,
            status: AssetStatus::Placeholder,
        }
    }

    /// Creates a linked asset reference pointing to a real asset.
    ///
    /// Only the Asset Linker (Phase 7) should call this — not game systems.
    /// Used when importing existing projects that already have assets mapped.
    pub fn linked(id: impl Into<String>, asset_type: AssetType) -> Self {
        Self {
            id: id.into(),
            asset_type,
            status: AssetStatus::Linked,
        }
    }

    /// Creates a sentinel "none" reference for fields that have no asset.
    ///
    /// Used for optional asset fields like `material_ref` in RenderComponent
    /// when no material override is needed. Distinct from Placeholder —
    /// "none" means intentionally empty, not "pending a real asset."
    pub fn none() -> Self {
        Self {
            id: "none".into(),
            asset_type: AssetType::Mesh,
            status: AssetStatus::Placeholder,
        }
    }

    /// Returns true if this reference is safe to commit to CGS.
    ///
    /// Only UNRESOLVED references block commit (I12).
    /// Placeholder, Linked, and Missing are all committable —
    /// they represent known states in the asset pipeline.
    pub fn is_committable(&self) -> bool {
        !matches!(self.status, AssetStatus::Unresolved)
    }

    /// Returns true if the engine can render this asset right now.
    /// Only Linked references have a real file the engine can load.
    pub fn is_renderable(&self) -> bool {
        matches!(self.status, AssetStatus::Linked)
    }

    /// Returns true if this reference is in Placeholder state.
    /// Placeholder = auto-created, no real file yet, game logic works.
    pub fn is_placeholder(&self) -> bool {
        matches!(self.status, AssetStatus::Placeholder)
    }

    /// Returns true if this reference is missing its linked file.
    /// Missing = was linked, file no longer found. Warning, not blocker.
    pub fn is_missing(&self) -> bool {
        matches!(self.status, AssetStatus::Missing)
    }

    /// Returns true if this reference is unresolved.
    /// Unresolved = referenced in schema but never registered. Bug. Blocks commit.
    pub fn is_unresolved(&self) -> bool {
        matches!(self.status, AssetStatus::Unresolved)
    }

    /// Transitions this reference to Linked state.
    /// Called by the Asset Linker when a real file is mapped to this ID.
    pub fn mark_linked(&mut self) {
        self.status = AssetStatus::Linked;
    }

    /// Transitions this reference to Missing state.
    /// Called by the Asset Validator when a previously linked file is not found.
    pub fn mark_missing(&mut self) {
        self.status = AssetStatus::Missing;
    }

    /// Transitions this reference to Unresolved state.
    /// Called by the Asset Validator when a reference is found in the schema
    /// but has never been registered in the asset pipeline.
    pub fn mark_unresolved(&mut self) {
        self.status = AssetStatus::Unresolved;
    }
}

impl std::fmt::Display for AssetReference {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}[{:?}]({:?})", self.id, self.asset_type, self.status)
    }
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn placeholder_starts_as_placeholder() {
        let r = AssetReference::placeholder(
            "character_knight_mesh_v1",
            AssetType::Mesh
        );
        assert!(r.is_placeholder());
        assert!(!r.is_renderable());
        assert!(r.is_committable());
        assert!(!r.is_unresolved());
    }

    #[test]
    fn linked_reference_is_renderable() {
        let r = AssetReference::linked("character_knight_mesh_v1", AssetType::Mesh);
        assert!(r.is_renderable());
        assert!(r.is_committable());
        assert!(!r.is_placeholder());
    }

    #[test]
    fn unresolved_blocks_commit() {
        let mut r = AssetReference::placeholder("bad_ref", AssetType::Texture);
        r.mark_unresolved();
        assert!(r.is_unresolved());
        assert!(!r.is_committable());
    }

    #[test]
    fn missing_is_committable_not_renderable() {
        let mut r = AssetReference::linked("old_asset_v1", AssetType::AudioClip);
        r.mark_missing();
        assert!(r.is_missing());
        assert!(r.is_committable());
        assert!(!r.is_renderable());
    }

    #[test]
    fn mark_linked_transitions_from_placeholder() {
        let mut r = AssetReference::placeholder("mesh_v1", AssetType::Mesh);
        r.mark_linked();
        assert!(r.is_renderable());
        assert!(!r.is_placeholder());
    }

    #[test]
    fn none_reference_is_committable() {
        let r = AssetReference::none();
        assert!(r.is_committable());
        assert_eq!(r.id, "none");
    }

    #[test]
    fn display_includes_id_and_status() {
        let r = AssetReference::placeholder("knight_mesh_v1", AssetType::Mesh);
        let display = r.to_string();
        assert!(display.contains("knight_mesh_v1"));
    }

    #[test]
    fn asset_type_preserved() {
        let r = AssetReference::placeholder("sfx_v1", AssetType::AudioClip);
        assert_eq!(r.asset_type, AssetType::AudioClip);
    }
}

File: packages/asset-registry/asset_reference.py
"""
asset_reference.py — Python-side AssetReference struct (Audit 2).

Mirrors packages/core/src/assets/asset_reference.rs exactly.
Every asset reference in the CGS and Asset Registry uses this struct —
NEVER a raw string.

## Why Typed References (Audit 2)
Raw strings like "character_knight_mesh_v1" cannot be validated, tracked,
or status-checked. A typed AssetReference carries:
  - A unique asset_id (the canonical name, never changes after assignment)
  - The AssetType (what kind of asset this is)
  - The AssetStatus (PLACEHOLDER / LINKED / MISSING / UNRESOLVED)
  - Optionally, the resolved_path (the engine-side file path when LINKED)

The Asset Registry holds the authoritative AssetReference for each asset_id.
The CGS stores only the asset_id string — the full reference is resolved
through the registry at validation time.

## Serialization
AssetReference serializes to/from a JSON dict with four keys:
  {"asset_id": str, "asset_type": str, "status": str, "resolved_path": str|null}

This is the format used in CGS snapshots and feedback payloads.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from asset_type_enum import AssetType
from asset_status_enum import AssetStatus


@dataclass
class AssetReference:
    """
    A typed, trackable reference to an engine asset.

    ## Immutable Fields
    asset_id and asset_type are set at creation and never change.
    Status and resolved_path are mutable — they change as the asset
    moves through its lifecycle.

    ## Naming Convention (Audit 2)
    asset_id follows the pattern: [entity_type]_[entity_name]_[asset_type]_v[N]
    Examples: character_knight_mesh_v1, enemy_dragon_roar_sfx_v1
    Use AssetNamingPolicy.generate() to produce canonical IDs.
    """

    # Unique identifier — follows naming convention, never reused
    asset_id: str

    # What kind of asset this is — immutable after creation
    asset_type: AssetType

    # Current lifecycle state
    status: AssetStatus = field(default=AssetStatus.PLACEHOLDER)

    # Engine-side file path — only populated when status == LINKED
    resolved_path: Optional[str] = field(default=None)

    # ── Validation ────────────────────────────────────────────────────────

    def __post_init__(self) -> None:
        if not self.asset_id:
            raise ValueError("AssetReference: asset_id must not be empty")
        if not self.asset_id.replace("_", "").replace("-", "").replace(".", "").isalnum():
            raise ValueError(
                f"AssetReference: asset_id '{self.asset_id}' contains invalid characters. "
                "Use only letters, digits, underscores, hyphens, and dots."
            )
        if self.status == AssetStatus.LINKED and not self.resolved_path:
            raise ValueError(
                f"AssetReference '{self.asset_id}': status is LINKED but resolved_path is None. "
                "A LINKED asset must have a resolved_path."
            )
        if self.status != AssetStatus.LINKED and self.resolved_path is not None:
            raise ValueError(
                f"AssetReference '{self.asset_id}': resolved_path is set but status is "
                f"{self.status.value}. Only LINKED assets carry a resolved_path."
            )

    # ── Status Transitions ────────────────────────────────────────────────

    def link(self, resolved_path: str) -> None:
        """
        Transitions this reference from PLACEHOLDER or MISSING to LINKED.

        Raises ValueError if:
        - resolved_path is empty
        - status is UNRESOLVED (cannot be resolved this way — it's a bug)
        """
        if not resolved_path:
            raise ValueError(
                f"AssetReference.link(): resolved_path must not be empty "
                f"for asset '{self.asset_id}'"
            )
        if self.status == AssetStatus.UNRESOLVED:
            raise ValueError(
                f"AssetReference '{self.asset_id}' is UNRESOLVED — it was never "
                "registered. Fix the registration bug before linking."
            )
        self.status = AssetStatus.LINKED
        self.resolved_path = resolved_path

    def mark_missing(self) -> None:
        """
        Transitions this reference from LINKED to MISSING.
        Called when the linked file can no longer be found.
        """
        if self.status != AssetStatus.LINKED:
            raise ValueError(
                f"AssetReference.mark_missing(): asset '{self.asset_id}' "
                f"is not LINKED (current status: {self.status.value})"
            )
        self.status = AssetStatus.MISSING
        self.resolved_path = None

    def revert_to_placeholder(self) -> None:
        """
        Resets this reference back to PLACEHOLDER.
        Used when an asset link is intentionally removed.
        """
        self.status = AssetStatus.PLACEHOLDER
        self.resolved_path = None

    # ── Classification ────────────────────────────────────────────────────

    @property
    def is_placeholder(self) -> bool:
        return self.status == AssetStatus.PLACEHOLDER

    @property
    def is_linked(self) -> bool:
        return self.status == AssetStatus.LINKED

    @property
    def is_missing(self) -> bool:
        return self.status == AssetStatus.MISSING

    @property
    def is_unresolved(self) -> bool:
        return self.status == AssetStatus.UNRESOLVED

    @property
    def blocks_cgs_commit(self) -> bool:
        """True if this reference blocks CGS commit (I12). Only UNRESOLVED does."""
        return self.status.blocks_cgs_commit

    @property
    def is_renderable(self) -> bool:
        """True if the engine can render/play this asset."""
        return self.status.is_renderable

    # ── Serialization ─────────────────────────────────────────────────────

    def to_dict(self) -> dict:
        """
        Serializes to a JSON-compatible dict.
        Format: {"asset_id": str, "asset_type": str, "status": str, "resolved_path": str|null}
        """
        return {
            "asset_id": self.asset_id,
            "asset_type": self.asset_type.value,
            "status": self.status.value,
            "resolved_path": self.resolved_path,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "AssetReference":
        """
        Deserializes from a JSON-compatible dict.
        Raises ValueError if required fields are missing or invalid.
        """
        required = {"asset_id", "asset_type", "status"}
        missing = required - data.keys()
        if missing:
            raise ValueError(
                f"AssetReference.from_dict(): missing required fields: {missing}"
            )
        return cls(
            asset_id=data["asset_id"],
            asset_type=AssetType.from_string(data["asset_type"]),
            status=AssetStatus.from_string(data["status"]),
            resolved_path=data.get("resolved_path"),
        )

    @classmethod
    def make_placeholder(cls, asset_id: str, asset_type: AssetType) -> "AssetReference":
        """
        Creates a new PLACEHOLDER AssetReference.
        This is the correct factory for auto-registration (Audit 2 zero-experience flow).
        """
        return cls(asset_id=asset_id, asset_type=asset_type, status=AssetStatus.PLACEHOLDER)

    @classmethod
    def make_unresolved(cls, asset_id: str, asset_type: AssetType) -> "AssetReference":
        """
        Creates an UNRESOLVED AssetReference.
        Only used when a CGS reference is detected that was never registered.
        """
        return cls(asset_id=asset_id, asset_type=asset_type, status=AssetStatus.UNRESOLVED)

    def __repr__(self) -> str:
        path_part = f", path='{self.resolved_path}'" if self.resolved_path else ""
        return (
            f"AssetReference(id='{self.asset_id}', "
            f"type={self.asset_type.value}, "
            f"status={self.status.value}{path_part})"
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, AssetReference):
            return NotImplemented
        return self.asset_id == other.asset_id

    def __hash__(self) -> int:
        return hash(self.asset_id)